### ISSUE TITLE
sappers and siege towers

### ISSUE THREAD

**Issue by ilteroi** [LoneGazebo/Community-Patch-DLL#4301]
today i learned that there is similar but not identical code for "sapper" units and "city attack support units".

can anyone tell me the difference and whether we really need both?

**Comment by Iamblichos** [comment]
Yeah we need both. Going back to vanilla. IsCityAttackSupport -> Battering Ram and Siege Tower -> Can Attack Only Cities. Sapper-> Siege Tower Only -> Area effect giving units at City additional city attack. In VP, the Siege Tower is reduced to a civilian unit it still has move instructions via IsCityAttackSupport, but it can't attack. Just gives out the Sapper ability or Half Sapper if it is within two tiles of City.

**Comment by Iamblichos** [comment]
BTW--sorry I didn't get back to you on this yesterday. Another way to look at it is that Sapper is a promotion only so it can be handed out to any unit, not just CityAttackSupport. For CityAttackSupport (Table entry CityAttackOnly), has both a UnitClassess entry and a Promotion class entry, either one of those will reduce any unit to only being able to attack cities. It's only tangentially related to Sapper, as IsNearCityAttackSupport uses the Sapper range Tactical AI and Homeland AI moves, but other than that, it has nothing to do with the Sapper ability.

**Comment by ilteroi** [comment]
so a siege tower has the sapper promotion anyway?

i´m just asking because the IsNearSapper check is a performance problem, and the simplest solution is replacing it with an IsNearCityAttackSupport check.

**Comment by Iamblichos** [comment]
The Siege Tower is the only unit I know of that has the Sapper promotion. If you add the Sapper check into IsNearCityAttackSupport then it would entail that Sapper would be dependent on having CityAttackOnly as the units ability as well. I don't see the problem except that they are really independent and could break mod or future mod compatibility. You would still have to distinguish them fully in IsNearCityAttackSupport since the battering ram doesn't do Sapper.

**Comment by Iamblichos** [comment]
Question for you: how is this any more expensive than IsNearGreatGeneral? Their codes are nearly identical in structure and call routine. Or is more than one call like that too muchin generating max strength and maxed ranged ?strength?

**Comment by ilteroi** [comment]
well, it's a loop over 18 plots for every call to getAttackStrength which in turn is called very often during combat simulation.

**Comment by Iamblichos** [comment]
We should really combine these since they are all common code we can search for any of the units after we do just one iteration of plots.

**Comment by Iamblichos** [comment]
Changes will be coming to this. I've combined IsNearSapper, IsNearCityAttackSupport, IsNearGreatGeneral which feeds into one common piece of code call. @ilteroi was finding very poor performance with the IsNearSapper....We will no longer need IsHalfSapper and all that or IsSappingCity, etc.

**Comment by Iamblichos** [comment]
One caveat I'm using int iGreatGeneralRange = /*2*/ GC.getGREAT_GENERAL_MAX_RANGE(); to divvy up the distance since the default sapper range is the same. So now they go hand in hand pretty much to make a common code.

**Comment by Iamblichos** [comment]
@ilteroi 

New branch:
https://github.com/LoneGazebo/Community-Patch-DLL/commit/c2661d03c1e38bfe62f21bfa123cd1aefc725137

**Comment by Iamblichos** [comment]
Make sure you check my commit in there, you'll have to grab updated table for something not relevant to this and get the new EnemyUnitPanel.lua for whatever version you need.

Test it out and let me know where we can improve.

**Comment by ilteroi** [comment]
just took a look ... just one comment:
```
	int iAuraEffectChange = 0;
	if (!bIgnoreUnitAdjacencyBoni && !IsIgnoreGreatGeneralBenefit())
	{
		iTempModifier = GetAreaEffectBonus(iAuraEffectChange, pBattlePlot, NULL, NULL, true, false, false);
		iModifier += iTempModifier;
		iModifier += iAuraEffectChange;
	}
```
seems strange, how about this

```
	if (!bIgnoreUnitAdjacencyBoni && !IsIgnoreGreatGeneralBenefit())
	{
		iModifier += GetAreaEffectBonus(pBattlePlot, NULL, NULL, true, false, false);
	}
```

otherwise it looks ok.

**Comment by Iamblichos** [comment]
We’ll be skipping a variable there. Trying to match the intent of whatever that fella who wrote the aura effect change stuff.

**Comment by Iamblichos** [comment]
I see what you are saying though I’ve already accounted for the aura effect bonus in the main function—if it exists. Don’t need to add it again in max strength.

**Comment by Iamblichos** [comment]
@ilteroi I finished the Sapper, General, and CityAttackSupport, and so I moved on to the IsNearbyPromotion type unit as we discussed. Instead of setting a promotion and check for a unit around in SetXY, I've now created a new player variable that works like the general m_unitsAreaEffectPromotion and created several bonus type functions to look for this unit and apply the bonus to appropriate units, should work better than doing that stuff in SetXY. At any rate, I need help with one function that I'd like to apply visibility rules to:

```
bool CvUnit::IsGiveInvisibilityToUnit(const CvPlot* pAtPlot) const
{
	VALIDATE_OBJECT
	int iRange = 0;
	if (pAtPlot == NULL)
	{
		pAtPlot = plot();
		if (pAtPlot == NULL)
			return false;
	}

	const std::vector<std::pair<int, int>>& possibleUnits = GET_PLAYER(getOwner()).GetAreaEffectPromotionUnits();
	for (std::vector<std::pair<int, int>>::const_iterator it = possibleUnits.begin(); it != possibleUnits.end(); ++it)
	{
		CvUnit* pUnit = GET_PLAYER(getOwner()).getUnit(it->first);
		if (pUnit && pUnit->isGiveInvisibility())
		{
			iRange = pUnit->GetNearbyUnitPromotionsRange();
			if (plotDistance(pAtPlot->getX(), pAtPlot->getY(), pUnit->getX(), pUnit->getY()) <= iRange)
			{
				if (pUnit->IsGiveDomainBonus(getDomainType()) != NULL && IsCombatUnit())
				{
					return true;
				}
			}
		}
	}
	return false;
}
```

Basically I'd like change the visibility of units that are within the defined iRange of pUnit, like an InvisibleType (without having to set these units as InvisibleType), and then reveal them if they aren't within range and/or if a neutral/enemy is adjacent.

**Comment by Iamblichos** [comment]
You can see the full code in my branch.

**Comment by Iamblichos** [comment]
These are my thoughts so far:

1) update IsInvisible and SetXY with appropriate code in CvUnit. Fairly simple and straight forward. 

2) Create a new variable in CvPlot to track the visible/invisible plots analogous to InvisibleTypes and have that updated in the appropriate places where changeInvisibleVisible is called.

In CvPlot such convoluted code.

**Comment by LoneGazebo** [comment]
Iamblichos be careful with adding parameters to CvPlot, that’s very expensive memory wise and we are quite close to our memory cap. I’d recommend against it.

**Comment by Iamblichos** [comment]
@LoneGazebo, it should only cache the plots if they get set by that code in CvUnit above. Will do some testing and see where I'm at. I run a little heavier than full VP anyway. If I'm good here, we'll be good there.

**Comment by Iamblichos** [comment]
Just to give you an idea, previously I was playing with two InvisibilityTypes with a promotion for it getting set in SetXY via DoNearbyUnitPromotion with m_apaiInvisibleVisibilityCount doing double duty. The code was quite inefficient, but huge map games did not run out of memory.

**Comment by LoneGazebo** [comment]
Whether or not the cache is filled/used the memory is initialized and set aside. That can hurt our memory profile.

**Comment by Iamblichos** [comment]
As I said, I was already using more than one NUM_INVISIBLE_TYPES, which from the vanilla code writer was a big no-no. I found that to be well...not much of a hit (if any). I'll be doing some testing, and if I find it to be too much of a performance hit, I won't make my next commit (code for the intent of @ilteroi is done in my current branch). If it's somewhat significant, or iffy, we just won't pull it into master branch. Or I can just make this next commit a global define, let people decide if they ever want to use it or not. Sound good?

**Comment by LoneGazebo** [comment]
Sounds fine to me. Can the visible status of a unit not be managed through a unit’s status?

**Comment by Iamblichos** [comment]
I have it working. Yeah technically it can in the CvUnit::isInvisible. Problem is that the plot visibility won't change unless you track changevisibilitycount of the plot. I find the whole method of the way they wrote the plot visibility to be asinine. I think @ilteroi shares the same headache.

**Comment by Iamblichos** [comment]
Not only must you update the plot visible/invisible, you have to update the graphics in SetXY like this:
```
if (eOurTeam != activeTeam && IsGiveInvisibilityToUnit(activeTeam))
		{
			bool bOldInvisibleVisible = false;
			if (pOldPlot)
				bOldInvisibleVisible = pOldPlot->isInvisibleVisibleUnit(activeTeam);
			bool bNewInvisibleVisible = pNewPlot->isInvisibleVisibleUnit(activeTeam);
			if (bOldInvisibleVisible != bNewInvisibleVisible)
			{
				auto_ptr<ICvUnit1> pDllUnit(new CvDllUnit(this));
				gDLL->GameplayUnitVisibility(pDllUnit.get(), bNewInvisibleVisible, true);
			}
		}
```

**Comment by LoneGazebo** [comment]
@iamblichos I get that, but the memory cost of tying it to units instead of plots is worth the complexity. Just write a subfunction for CvUnit that automates it for you after you change an invisibility state.

**Comment by Iamblichos** [comment]
@LoneGazebo, I'm not sure that would work. While I'm sure you can be successful in updating the unit state which is fairly straight forward, if the state of the plot--just the way they have set it all up with plot visibility--is unchanged, then I'm not sure that the AI will "see" what you see given identical perspectives. Let me think about it some more.

**Comment by Iamblichos** [comment]
The whole idea of this IsGiveInvisilityToUnit was inspired by the movie The Last Samurai (one of my favorites) where Capt. Nathan Algren first encountered the Samurai and they were hiding "in the fog" and waiting to ambush the conscripted Imperial Soldiers. Feature is fully working.

![20180507075014_1](https://user-images.githubusercontent.com/14372214/39702922-fb9278f6-51cb-11e8-857e-f2907e8be13c.jpg)

![20180507075109_1](https://user-images.githubusercontent.com/14372214/39702926-ffcdb9a8-51cb-11e8-8e0c-1c5098129c06.jpg)

My units are Denmark and I gave Belgrade a Japanese Daimyo (unique General).

**Comment by LoneGazebo** [comment]
@iamblichos I use very sleepy. Install it, run it, hook it into civ when civ is running and auto play. Let it run for a few hundred turns then stop it and look at the inclusive performance.

**Comment by Iamblichos** [comment]
@LoneGazebo, excellent, thank you.

**Comment by Iamblichos** [comment]
@LoneGazebo CvUnit::IsNearSapper and CvUnit::IsHalfNearSapper will no longer be called anywhere, nor their respecting CvLuaUnit calls in EnemyUnitPanel.. CvLuaUnit has consolidated those too into a new function. Do you want me to delete the old ones?

**Comment by ilteroi** [comment]
two comments:

* i find the name "IsGiveInvisibilityToUnit" a bit weird, how about "IsHiddenByNearbyUnit"?

* you can choose the thread to profile in verysleepy. that cuts down on the noise in the result. it's a bit tricky to pick the right thread though, because it's cpu load varies a lot. so i usually sort and re-sort by cpu load descending until the callstack for the topmost thread indicates some familiar DLL functions being called.

**Comment by Iamblichos** [comment]
- Sure I'll change the name

- I'll see if I can sort through it.

**Comment by Iamblichos** [comment]
@ilteroi instead of using the IsHalfSappingCity call in GetAreaEffectBonus which iterates over all the plots within distance of pUnit->IsSapper isn't much cheaper just to do this since we know the plot positions:

```
else if (pTargetCity->plot()->isEnemyCity(*pUnit) && (plotDistance(pUnit->getX(), pUnit->getY(), pTargetCity->getX(), pTargetCity->getY()) == iGreatGeneralRange))
						{
							iBonus = GC.getSAPPED_CITY_ATTACK_MODIFIER() / 2;
						}
					}
```

**Comment by Iamblichos** [comment]
@ilteroi @LoneGazebo 

```
else if (bSapper && !bGreatGeneral && !bCityAttackSupport && pUnit->IsSapper() && pUnit != this)
				{
					CvAssertMsg(pTargetCity, "Target city is NULL when checking sapping combat bonus");
					if (pTargetCity != NULL)
					{
						if (pUnit->IsEnemyCityAdjacent(pTargetCity))
						{
							iBonus = GC.getSAPPED_CITY_ATTACK_MODIFIER();
						}
						else if (pTargetCity->plot()->isEnemyCity(*pUnit) && (plotDistance(pUnit->getX(), pUnit->getY(), pTargetCity->getX(), pTargetCity->getY()) == iGreatGeneralRange))
						{
							iBonus = GC.getSAPPED_CITY_ATTACK_MODIFIER() / 2;
						}
					}
				}
```

I'm adding in pUnit != this so the siege tower doesn't give a bonus to itself if it's not supported by another siege tower (assuming the vanilla version with a combat rating). That's its intent right?

**Comment by ilteroi** [comment]
well i regard that whole half sapper thing as dispensable ... so don't ask me :)

but the rest looks good.

**Comment by Iamblichos** [comment]
So you had two things going on with Sappers...IsNearSapper iterates plots around a unit to find a sapper, and within that you have IsSappingCity which isn't a big deal since that just looks at the first ring. Half Sapper is the worst, iterate around plots to find a Sapper, and then IsHalfSappingCity iterates around the Sapper--within a plot range of 2--to try and find an enemy city. I've now dispensed with all four of those functions. IsSapper is an m_unitsAreaEffectPositive, so we always know its plot position. The unit plot and the city plot are a given, so we just do plot distance check as above and away we go. Done.

**Comment by Iamblichos** [comment]
Finished up my testing. Everything is working great. My main function call in CvPlot, rarely got called. IsHiddenByNearbyUnit in CvUnit got called quite a bit, but not nearly as much as IsHoveringUnit....there's just enough of those generals to really effect make a huge impact to the CvUnit calls. I have additional checks in CvUnit init, kill, and SetXY if the general moves, is killed, or is initialized but those don't get called that much, especially for a great person.

**Comment by Iamblichos** [comment]
@LoneGazebo all changes are in my branch. I can do a pullrequest to main branch or you can sync up to mine. Mine is ahead of all branches. I'm also sync'ed up with shiny and mp_bugfixes. Dunno if that'll be a problem with a pull request or not, but there shouldn't be any conflicts.
